### PR TITLE
feat(contact-center): added-mpc

### PR DESCRIPTION
**Pull request by adhmenon**
# COMPLETES #< [CAI-7286](https://jira-eng-sjc12.cisco.com/jira/browse/CAI-7286)>

## This pull request addresses

* Adds most features of the MPC (Multi party conference)
* Fixed and updated unit tests of consult flow

## by making the following changes

* Added code in state machine and actions for MPC.
* Added sample app changes.

Vidcast - https://app.vidcast.io/share/ce3d2942-79d9-499f-9678-90ca4dda9d13?playerMode=vidcast

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios were tested

* Tested manually as per the widgets test spec.
* Tested 3PC and MPC only.
* Some issues remain like agent list & merge - will be handled in future PRs

## The GAI Coding Policy And Copyright Annotation Best Practices ##
 

 
- [ ] GAI was not used (or, no additional notation is required)
- [x] Code was generated entirely by GAI
- [ ] GAI was used to create a draft that was subsequently customized or modified
- [ ] Coder created a draft manually that was non-substantively modified by GAI (e.g., refactoring was performed by GAI on manually written code)
- [ ] Tool used for AI assistance (GitHub Copilot / Other - specify)
  - [ ] Github Copilot
  - [x] Cursor
- [ ] This PR is related to
  - [ ] Feature
  - [ ] Defect fix
  - [x] Tech Debt
  - [ ] Automation

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request
- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

**Review by adhmenon — COMMENTED** [review]
Some flags are redundant, working on removing them and optimising the guard cases.

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 1]
```diff
@@ -2,6 +2,280 @@
...
+  MEDIA_TYPE_MAIN_CALL,
+} from './state-machine/constants';
+
+// Re-export for backward compatibility
+export {MAX_PARTICIPANTS_IN_MULTIPARTY_CONFERENCE};
+
+/**
+ * Media entry type from interaction.media
+ */
+type MediaEntry = {
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts]: Shouldn't this and below be present in types file ?

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 2]
```diff
@@ -2,6 +2,280 @@
...
+  return {isHeld: mediaEntry.isHold === true, mediaResourceId};
+};
+
+/**
+ * Gets the hold status of the consult call leg
+ * @param interaction - The interaction object
+ * @param interactionId - The main interaction ID
+ * @returns true if consult call is held
+ */
+export const getConsultCallHeld = (interaction: Interaction, interactionId: string): boolean => {
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts]: I feel these 2 methods are unncessary because anyway when to pass the the Media type as consult and media type as main call is decided based on where you call this method so directly we can call findHoldStatus() from those places and pass the right media type

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 3]
```diff
@@ -2,6 +2,280 @@
...
+ * Checks if the agent is on the consult leg and the consult is NOT held
+ * (meaning the main call is effectively held from the agent's perspective)
+ * Per widgets-util-logic.md: isConsultOnHoldMPC implementation
+ *
+ * @param interaction - The interaction object
+ * @param interactionId - The main interaction ID (kept for API compatibility)
+ * @returns true if agent is on consult (and thus main call is held from their POV)
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+export const isConsultOnHoldMPC = (interaction: Interaction, interactionId: string): boolean => {
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts]: We seem to be passing interactionId in all these methods but is isn't really getting used

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 4]
```diff
@@ -2,6 +2,280 @@
...
+  return findHoldStatus(interaction, interactionId, MEDIA_TYPE_CONSULT).isHeld;
+};
+
+/**
+ * Gets the hold status of the main call leg
+ * @param interaction - The interaction object
+ * @param interactionId - The main interaction ID
+ * @returns true if main call is held
+ */
+export const getMainCallHeld = (interaction: Interaction, interactionId: string): boolean => {
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts]: Above 4 methods can be combined into one single method and and return isHeld status directly based on per call leg

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 5]
```diff
@@ -2,6 +2,280 @@
...
+  return findHoldStatus(interaction, interactionId, MEDIA_TYPE_MAIN_CALL).isHeld;
+};
+
+/**
+ * Checks if the customer is still in the call (not left)
+ * Per widgets-util-logic.md: getIsCustomerInCall implementation
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts]: Please remove this comment from everywhere

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 6]
```diff
@@ -2,6 +2,280 @@
...
+ *
+ * @param interaction - The interaction object
+ * @param interactionId - The main interaction ID
+ * @returns true if customer is in the call
+ */
+export const getIsCustomerInCall = (interaction: Interaction, interactionId: string): boolean => {
+  if (!interaction?.media || !interaction?.participants) {
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts]: I believe these checks in all the utility method are inncessary because interaction seems to be a mandatory param and ideally we should not even be invoking these util method if interaction object itself can be optional

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 7]
```diff
@@ -2,6 +2,280 @@
...
+    return false;
+  }
+
+  // Get the main call media
+  const mainCallMedia = interaction.media[interactionId];
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts:16]: Why is this extra check needed ?

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 8]
```diff
@@ -2,6 +2,280 @@
...
+
+/**
+ * Checks if the maximum number of participants in a conference has been reached
+ * Per conference-spec.md: Max 7 counted participants + 1 customer
+ *
+ * @param interaction - The interaction object
+ * @param interactionId - The main interaction ID
+ * @returns true if max participants reached
+ */
+export const hasReachedMaxParticipants = (
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts]: This method is not adding much value

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 9]
```diff
@@ -2,6 +2,280 @@
...
+  const conferenceParticipantsCount = getConferenceParticipantsCount(interaction, interactionId);
+  const isConsultInProgress = getIsConsultInProgress(interaction);
+  const isCustomerInCall = getIsCustomerInCall(interaction, interactionId);
+
+  // Per widgets-util-logic.md gating logic:
+  // 1. Must not be at max participants
+  // 2. Must not have consult already in progress
+  // 3. Customer must be in call
+  // 4. Special case: if held AND conference in progress AND consult not completed, disable
+  if (conferenceParticipantsCount >= MAX_PARTICIPANTS_IN_MULTIPARTY_CONFERENCE) {
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts]: These ifs can be combined into one single condition

File: packages/@webex/contact-center/src/services/task/Task.ts [hunk 10]
```diff
@@ -71,7 +71,8 @@ export default abstract class Task extends EventEmitter implements ITask {
     super();
     this.contact = contact;
     this.data = data;
-    this.uiControlConfig = uiControlConfig;
+    // Include agentId in the config for ownership checks (transfer conference)
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/Task.ts:76]: We can remove this comment

File: packages/@webex/contact-center/src/services/task/state-machine/constants.ts [hunk 11]
```diff
@@ -3,6 +3,43 @@
...
 
+// ============================================
+// Conference Constants
+// ============================================
+
+/**
+ * Maximum number of participants allowed in a multi-party conference.
+ * Max 7 counted agents + 1 customer.
+ */
+export const MAX_PARTICIPANTS_IN_MULTIPARTY_CONFERENCE = 7;
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/constants.ts:14]: We can just name it MAX_PARTICIPANTS

File: packages/@webex/contact-center/src/services/task/state-machine/constants.ts [hunk 12]
```diff
@@ -3,6 +3,43 @@
...
+
+// ============================================
+// Participant Type Constants
+// ============================================
+
+/** Customer participant type */
+export const PARTICIPANT_TYPE_CUSTOMER = 'Customer';
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/constants.ts]: These are categories for the same thing so they can be created as type: PARTICIPANT_TYPE and in that you have mutliple values

File: packages/@webex/contact-center/src/services/task/state-machine/types.ts [hunk 13]
```diff
@@ -150,17 +133,36 @@ interface TaskEventPayloadMap {
...
   [TaskEvent.CONFERENCE_START]: BaseEvent<TaskEvent.CONFERENCE_START> & {
     participants?: ConferenceParticipant[];
+    taskData?: TaskData;
   };
   [TaskEvent.CONFERENCE_FAILED]: BaseEvent<TaskEvent.CONFERENCE_FAILED> & {
     reason?: string;
+    taskData?: TaskData;
   };
-  [TaskEvent.CONFERENCE_END]: BaseEvent<TaskEvent.CONFERENCE_END>;
+  [TaskEvent.CONFERENCE_END]: BaseEvent<TaskEvent.CONFERENCE_END> & {taskData?: TaskData};
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/types.ts]: Why is taskData optional in these conference events?

File: packages/@webex/contact-center/src/services/task/state-machine/guards.ts [hunk 14]
```diff
@@ -4,35 +4,575 @@
...
+  },
+
+  // ============================================
+  // Consult Guards
+  // ============================================
+
+  /**
+   * Check if a consult is currently in progress
+   */
+  consultInProgress: ({context}: GuardParams): boolean => {
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/guards.ts]: We don't need guards for positive and negative both usecases, just one condition should suffice both. Let's discuss the guards created in this file if they are really needed and what all would be needed

**Review by Kesari3008 — CHANGES_REQUESTED** [review]

**Comment by adhmenon** [comment]
Tested everything - most cases work fine. Only switch and some minor UI issues are left.

File: packages/@webex/contact-center/src/services/task/voice/Voice.ts [hunk 15]
```diff
@@ -722,6 +747,270 @@ export default class Voice extends Task implements IVoice {
...
+      throw detailedError;
+    }
+  }
+
+  /**
+   * Check if conference is in progress from task data
+   * Uses same logic as guards.isConferenceInProgress but directly on instance data
+   * @returns true if 2+ agents are in the main call
+   */
+  private checkIsConferenceInProgress(): boolean {
```
Comment (adhmenon) [inline packages/@webex/contact-center/src/services/task/voice/Voice.ts]: Nit: Will be moving this to utils - have done the cleanup locally but will push after Arun's changes are merged and probably along with other review comments.

File: packages/@webex/contact-center/src/services/task/TaskManager.ts [hunk 16]
```diff
@@ -177,11 +177,13 @@ export default class TaskManager extends EventEmitter {
...
    * @param ccEvent - The CC_EVENT type from WebSocket
    * @param payload - The event payload
+   * @param agentId - Optional agent ID for state detection (needed for HYDRATE)
    * @returns TaskEventPayload for state machine or null if no mapping
    */
   private static mapEventToTaskStateMachineEvent(
     ccEvent: CC_EVENTS,
-    payload: WebSocketPayload
+    payload: WebSocketPayload,
+    agentId?: string
```
Comment (adhmenon) [inline packages/@webex/contact-center/src/services/task/TaskManager.ts:147]: Needed because otherwise some `AgentConsultEnded` events were not getting mapped properly - they were being cleared by the state machine during `clearConsultState`.

File: packages/@webex/contact-center/src/services/task/TaskManager.ts [hunk 17]
```diff
@@ -412,7 +439,40 @@ export default class TaskManager extends EventEmitter {
       return null;
     }
 
-    const task = this.taskCollection[message.data.interactionId];
+    // Some consult/conference events can arrive on a child interactionId while the task we track
+    // is keyed by the main/parent interactionId. Resolve to an existing task when possible.
+    const interactionId = message.data.interactionId;
```
Comment (adhmenon) [inline packages/@webex/contact-center/src/services/task/TaskManager.ts:386]: Nit: Will revert this - added by mistake.

File: packages/@webex/contact-center/src/services/task/TaskManager.ts [hunk 18]
```diff
@@ -247,16 +217,18 @@ export default class TaskManager extends EventEmitter {
...
           type: TaskEvent.TRANSFER_SUCCESS,
           taskData: payload,
         };
 
-      case CC_EVENTS.AGENT_VTEAM_TRANSFERRED:
       case CC_EVENTS.AGENT_WRAPUP:
+        return {type: TaskEvent.TASK_WRAPUP, taskData: {...payload, wrapUpRequired: true}};
       case CC_EVENTS.AGENT_CONTACT_UNASSIGNED:
-        return {type: TaskEvent.END, taskData: {...payload, wrapUpRequired: true}};
+        return null; // Add WRAPUP if needed
```
Comment (rsarika) [inline packages/@webex/contact-center/src/services/task/TaskManager.ts:232]: `AGENT_CONTACT_UNASSIGNED` now maps to `null`, so the state machine never sees a terminal/wrapup event. Previously this flowed to END with `wrapUpRequired: true`. If this event still means “agent unassigned from task”, skipping the state machine can leave the task in collection and skip cleanup/wrapup. Should this map to `TASK_WRAPUP` or `CONTACT_ENDED` instead of `null`?

File: packages/@webex/contact-center/src/services/task/state-machine/guards.ts [hunk 19]
```diff
@@ -4,46 +4,311 @@
...
+      ? getIsConferenceInProgress(context.taskData)
+      : false;
+
+    return conferenceInEvent || conferenceInContext;
+  },
+
+  shouldDowngradeConference: ({context, event}: GuardParams): boolean => {
+    const eventTaskData = getTaskDataFromEvent(event);
+    const taskData = eventTaskData ?? context.taskData;
+    if (!taskData?.interaction || !taskData?.interactionId) return false;
```
Comment (rsarika) [inline packages/@webex/contact-center/src/services/task/state-machine/guards.ts]: `shouldDowngradeConference()` uses `taskData.interactionId` when counting participants. For consulted agents, `interactionId` may differ from the main call; other helpers now use `mainInteractionId`. Should this use `taskData.interaction.mainInteractionId || taskData.interactionId` to avoid miscounting and incorrect downgrade/termination?

**Review by rsarika — COMMENTED** [review]
Posting inline comments based on verified diffs.

File: packages/@webex/contact-center/src/services/task/state-machine/guards.ts (hunk 19 above)
Comment (adhmenon) [reply packages/@webex/contact-center/src/services/task/state-machine/guards.ts]: This is valid, however I was using this before but felt it might not have been needed.
As of now, things do seem to be working with the current interactionId since the rest of our guards are adequate here.

File: packages/@webex/contact-center/src/services/task/TaskManager.ts (hunk 18 above)
Comment (adhmenon) [reply packages/@webex/contact-center/src/services/task/TaskManager.ts:232]: This should be fine for now Ravi, since backend will always send `AGENT_WRAPUP` after this event, if needed. I believe in case of queue consult-transfer this happens.
In that case our state machine moves to `WRAPPING_UP`.
If not, we just go to `TERMINATED`. But we can test these edge cases, if needed. Have already tested it out.

**Comment by adhmenon** [comment]
Full MPC + Edge Cases - https://app.vidcast.io/share/d66170ec-06cd-4e21-9279-dcbcd5358613

MPC + Downgrade - https://app.vidcast.io/share/19a105d2-5c7a-4f48-8bf0-68739d1f7cb3

File: packages/@webex/contact-center/src/services/task/state-machine/constants.ts [hunk 20]
```diff
@@ -3,10 +3,46 @@
...
+
+// ============================================
+// Participant Type Constants
+// ============================================
+
+/** Participant types for conference filtering */
+export const PARTICIPANT_TYPE = {
+  CUSTOMER: 'Customer',
+  SUPERVISOR: 'Supervisor',
+  VVA: 'VVA',
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/constants.ts:24]: Which participant type is this? Are we using it ?

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 21]
```diff
@@ -597,74 +550,239 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
...
-          // AgentConferenceEnded
-          [TaskEvent.CONFERENCE_END]: {
-            target: TaskState.CONNECTED,
-            actions: ['clearConsultState'],
+          [TaskEvent.EXIT_CONFERENCE_FAILED]: {
+            actions: ['handleExitConferenceFailed'],
+          },
+
+          [TaskEvent.TRANSFER_CONFERENCE]: {
+            actions: ['emitTaskTransferConference'],
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts]: If we are emitting the event towards client on this transition itself then what value Transfer_ConferenceSuccess is adding in state machine

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 22]
```diff
@@ -597,74 +550,239 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
...
+                'emitTaskParticipantLeft',
+                'emitTaskConferenceEnded',
+              ],
             },
+            {actions: ['updateTaskData', 'handleParticipantLeft', 'emitTaskParticipantLeft']},
           ],
-          // AgentConferenceTransferred
-          [TaskEvent.TRANSFER_CONFERENCE]: [
+          [TaskEvent.EXIT_CONFERENCE]: {
+            actions: ['setExitingConference', 'emitTaskExitConference'],
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts]: Same question here

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 23]
```diff
@@ -597,74 +550,239 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
...
+                'updateTaskData',
+                'markEnded',
+                'clearConsultState',
+                'handleExitConferenceSuccess',
+                'emitTaskWrapup',
+              ],
+            },
+            {
+              // Agent exited → TERMINATED
+              guard: guards.isExitingConference,
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts]: Why is this guard needed? What is it helping in terms of decision making?

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 24]
```diff
@@ -597,74 +550,239 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
...
+          [TaskEvent.CONFERENCE_END]: [
+            {
+              // Owner/primary who should wrap up → WRAPPING_UP (must be first!)
+              guard: guards.shouldWrapUp,
+              target: TaskState.WRAPPING_UP,
+              actions: ['updateTaskData', 'markEnded', 'clearConsultState', 'emitTaskWrapup'],
+            },
+            {
+              // Non-owner who triggered exit → TERMINATED
+              guard: guards.isExitingConference,
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts]: There is no difference in action here and the one without guard down, transition state is also Terminated for both so why is the guard needed at all and why 2 blocks. Single section for it without the guard should suffice

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 25]
```diff
@@ -417,39 +321,60 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
...
             actions: ['updateTaskData', 'handleConsultFailed'],
           },
-          // AgentConsultCreated
-          [TaskEvent.CONSULT_CREATED]: {
-            target: TaskState.CONSULTING,
-            actions: ['updateTaskData', 'setConsultInitiator', 'emitTaskConsultCreated'],
-          },
+          [TaskEvent.CONSULT_CREATED]: [
+            {
+              guard: guards.isConsultingAgentOrBeingConsulted,
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts]: This guard should not be required since Consult Created is only received by the initiator and also user will only be in CONSULT_Initiating state who actually pressed the consult button so this transition can happen without the guard

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 26]
```diff
@@ -417,39 +321,60 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
...
             {
-              guard: isConsultQueueFlow,
+              guard: guards.isConsultQueueFlow,
               target: TaskState.CONNECTED,
               actions: ['updateTaskData', 'handleConsultFailed'],
             },
             {
-              guard: serverReportsHeld,
+              // If this consult originated from conference, always return to CONFERENCING.
+              guard: ({context}) => context.consultFromConference === true,
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts:342]: If we can use context directly here for decision making then all the one liner guadrs can be avoided and context can directly be used. Also why do we need to check on this and in all the transitions to Conferencing state. There is no dofference in target state or actions invoked so why are we using 3 different guards and doing it 3 times

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 27]
```diff
@@ -461,17 +386,17 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
           // AgentCtqCancelled Event
           [TaskEvent.CTQ_CANCEL]: [
             {
-              guard: isConsultQueueFlow,
+              guard: guards.isConsultQueueFlow,
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts]: This guard is also not needed

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 28]
```diff
@@ -417,39 +321,60 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
...
+              actions: ['updateTaskData', 'handleConsultFailed'],
+            },
+            {
+              guard: guards.serverReportsHeld,
               target: TaskState.HELD,
               actions: ['updateTaskData', 'handleConsultFailed'],
             },
             {
-              guard: serverReportsConsulting,
+              guard: guards.serverReportsConsulting,
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts]: If consult failure is received in consult initiating state, what case can be there where we would want to transition to consulting state. This doesn't seem to be a valid case

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 29]
```diff
@@ -2,6 +2,126 @@
 import {Interaction, ITask, TaskData, MEDIA_CHANNEL} from './types';
 import {OUTDIAL_DIRECTION, OUTDIAL_MEDIA_TYPE, OUTBOUND_TYPE} from '../../constants';
 import {LoginOption} from '../../types';
+import {PARTICIPANT_TYPE, MEDIA_TYPE_MAIN_CALL} from './state-machine/constants';
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts:5]: Now that we have action, guards, context and ui controls, how relevant this file is ? Do we need it ? Can util methods being used here not be combined with state machine logic ? Having both appraches are redundant

File: packages/@webex/contact-center/src/services/task/types.ts [hunk 30]
```diff
@@ -1811,3 +1849,44 @@ export interface IWebRTC extends IVoice {
...
+/**
+ * Context for processing an event
+ *
+ * Contains all information needed to process a WebSocket event:
+ * - Event type and payload from the backend
+ * - Task instance (if exists)
+ * - Pre-mapped state machine event (if applicable)
+ * - Task state flags (e.g., was this a consulted task)
+ */
+export interface EventContext {
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/types.ts:1874]: Having only this property: wasConsultedTask here in this type looks little odd as oher objects are bigger and high level objects ? Is there no other type to fit this in

File: packages/@webex/contact-center/src/services/task/types.ts [hunk 31]
```diff
@@ -1782,6 +1806,20 @@ export type Participant = {
  */
 export type TaskAccessorParticipant = Participant;
 
+/**
+ * Media entry type from interaction.media
+ * Used for media state tracking in consult and conference scenarios
+ */
+export type MediaEntry = {
```
Comment (Kesari3008) [inline packages/@webex/contact-center/src/services/task/types.ts]: Where is this getting used ? Couldn't find it getting used

**Review by Kesari3008 — CHANGES_REQUESTED** [review]
Let's discuss the statemahcine transitions, guards and action in the meeting for now. Rest of the files are mostly reviewed except tests

**Comment by adhmenon** [comment]
Tested after changes, everything works fine - https://app.vidcast.io/share/41b49e60-fcf8-4089-b630-82b8c5bd8b43

File: packages/@webex/contact-center/src/services/task/types.ts (hunk 30 above)
Comment (adhmenon) [reply packages/@webex/contact-center/src/services/task/types.ts:1874]: Addressed.

File: packages/@webex/contact-center/src/services/task/state-machine/constants.ts (hunk 20 above)
Comment (adhmenon) [reply packages/@webex/contact-center/src/services/task/state-machine/constants.ts:24]: Talked with Ravi, I think we should keep this for the virtual agent. Might be useful in the future.

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts (hunk 29 above)
Comment (adhmenon) [reply packages/@webex/contact-center/src/services/task/TaskUtils.ts:5]: I have combined as many utils as possible with the guards and ui controls.. did not want to go further but we can take it up as an effort with the next PR.

File: packages/@webex/contact-center/src/services/task/TaskManager.ts (hunk 18 above)
Comment (rsarika) [inline packages/@webex/contact-center/src/services/task/TaskManager.ts:232]: **Critical:** `AGENT_CONTACT_UNASSIGNED` previously mapped to `END` with `wrapUpRequired: true`. Now it returns `null`, so the state machine never sees a terminal event. If the backend doesn't always follow up with `AGENT_WRAPUP` or `CONTACT_ENDED`, the task will stay in the collection indefinitely with no cleanup.

Is there a guarantee from the backend that a follow-up event always arrives? If not, this should map to `TASK_WRAPUP` or `CONTACT_ENDED`.

File: packages/@webex/contact-center/src/services/task/state-machine/guards.ts [hunk 32]
```diff
@@ -4,46 +4,248 @@
...
+  /**
+   * Conference downgrade check specifically for transitioning back to CONNECTED.
+   *
+   * Returns true only when:
+   * - conference has downgraded (fewer than 2 active agent participants in main call)
+   * - customer is still in the call
+   * - current agent is still in the main call
+   */
+  shouldDowngradeConferenceToConnected: ({context, event}: GuardParams): boolean => {
+    const eventTaskData = getTaskDataFromEvent(event);
```
Comment (rsarika) [inline packages/@webex/contact-center/src/services/task/state-machine/guards.ts:146]: **Note:** For consulted agents, `taskData.interactionId` may differ from the main call's ID. If `mainInteractionId` is not populated and `interactionId` is a child interaction, `getConferenceParticipantsCount` will look up the wrong media entry, return 0, and incorrectly trigger a downgrade.

Consider logging a warning when `mainInteractionId` is missing so this edge case is surfaced during debugging.

File: packages/@webex/contact-center/src/services/task/voice/Voice.ts [hunk 33]
```diff
@@ -672,6 +693,13 @@ export default class Voice extends Task implements IVoice {
         data: paramsDataForConferenceV2.data,
       });
 
+      // Send success event to transition to CONFERENCING
+      if (this.stateMachineService) {
+        this.stateMachineService.send({
+          type: TaskEvent.CONFERENCE_START,
+        });
```
Comment (rsarika) [inline packages/@webex/contact-center/src/services/task/voice/Voice.ts:700]: **Minor:** `CONFERENCE_START` is sent locally here after the API resolves, but the backend will also send `AGENT_CONSULT_CONFERENCED` which maps to `CONFERENCE_START` in TaskManager (line 274). This may cause a duplicate conference-started emission to consumers.

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 34]
```diff
@@ -259,67 +190,50 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
...
-            actions: ['updateTaskData'],
           },
           // Click of the consult button
           [TaskEvent.CONSULT]: {
             target: TaskState.CONSULT_INITIATING,
             actions: ['setConsultInitiator', 'setConsultDestination'],
           },
-          // AgentConsultCreated event confirms the consult request
-          [TaskEvent.CONSULT_CREATED]: {
-            target: TaskState.CONSULTING,
```
Comment (rsarika) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts:279]: **Note:** `CONSULT_CREATED` handling was removed here, but `TaskManager.mapEventToTaskStateMachineEvent` still maps `AGENT_CONSULT_CREATED` → `TaskEvent.CONSULT_CREATED` (TaskManager line 194). The event is now silently dropped by the state machine. If this is intentional, consider also removing the mapping in TaskManager to keep things consistent.

File: packages/@webex/contact-center/src/services/task/Task.ts [hunk 35]
```diff
@@ -326,6 +402,9 @@ export default abstract class Task extends EventEmitter implements ITask {
 
   private getCommonActionOverrides(): Partial<TaskActionsMap> {
     return {
+      syncTaskDataFromEvent: ({event}: {event: TaskEventPayload}) => {
+        this.updateTaskFromEvent(event);
```
Comment (rsarika) [inline packages/@webex/contact-center/src/services/task/Task.ts:406]: **Minor:** `CONTACT_UPDATED` and `CONTACT_OWNER_CHANGED` now run both `updateTaskData` (updates XState context.taskData) and `syncTaskDataFromEvent` (updates Task instance `this.data`). This dual-path update works but is worth being aware of — if a future refactor splits these, the two data stores could diverge.

**Review by rsarika — APPROVED** [review]
Approving with inline comments on a few items to address or track. The MPC state machine design is solid — clean separation between TaskManager lifecycle and state machine transitions, well-organized guards, and thorough UI controls computation. Nice work on the refactoring.

**Comment by adhmenon** [comment]
Merging the PR now. Ravi, I'll handle the comments with the switch PR.
